File: backend/app/core/permissions.py
"""RBAC permission keys — align with frontend tab ids and admin capabilities."""

# Tool / navigation
PERM_DASHBOARD = "dashboard"
PERM_SERVICES = "services"
PERM_SMOKE = "smoke"
PERM_PATCHING = "patching"
PERM_K8 = "k8"
PERM_SSL = "ssl"
PERM_CONFIG_TREE = "configTree"
PERM_ALERTS = "alerts"

# Settings: appearance (theme) only — typical users
PERM_SETTINGS_APPEARANCE = "settings_appearance"
# Full settings: API keys, runtime, audit, legacy admin login, config registry
PERM_SETTINGS_FULL = "settings_full"
# Create/update/delete users and assign permissions
PERM_USERS_MANAGE = "users_manage"
# Add/update/delete monitored services, patching groups, K8 clusters in JSON
PERM_CONFIG_WRITE = "config_write"

ALL_TOOL_PERMISSIONS = [
    PERM_DASHBOARD,
    PERM_SERVICES,
    PERM_SMOKE,
    PERM_PATCHING,
    PERM_K8,
    PERM_SSL,
    PERM_CONFIG_TREE,
    PERM_ALERTS,
]

DEFAULT_USER_PERMISSIONS = ALL_TOOL_PERMISSIONS + [PERM_SETTINGS_APPEARANCE]

WILDCARD = "*"


def has_permission(role: str, permissions: list[str], required: str) -> bool:
    if role == "admin":
        return True
    if WILDCARD in permissions:
        return True
    return required in permissions


def has_any_tool(permissions: list[str], role: str) -> bool:
    if role == "admin" or WILDCARD in permissions:
        return True
    return any(p in permissions for p in ALL_TOOL_PERMISSIONS)

File: frontend/src/lib/permissions.js
/** Align with backend app/core/permissions.py */
export const PERM = {
  dashboard: "dashboard",
  services: "services",
  smoke: "smoke",
  patching: "patching",
  k8: "k8",
  ssl: "ssl",
  configTree: "configTree",
  alerts: "alerts",
  settingsAppearance: "settings_appearance",
  settingsFull: "settings_full",
  usersManage: "users_manage",
  configWrite: "config_write",
};

export const TAB_ORDER = [
  "dashboard",
  "services",
  "smoke",
  "patching",
  "k8",
  "ssl",
  "configTree",
  "alerts",
  "settings",
];

export function canAccessTab(user) {
  if (!user) return () => false;
  const { role, permissions = [] } = user;
  if (role === "admin" || permissions.includes("*")) return () => true;
  return (tabId) => {
    if (tabId === "settings") {
      return (
        permissions.includes(PERM.settingsAppearance) ||
        permissions.includes(PERM.settingsFull) ||
        permissions.includes(PERM.usersManage)
      );
    }
    return permissions.includes(tabId);
  };
}

export function firstAllowedTab(user) {
  const can = canAccessTab(user);
  for (const id of TAB_ORDER) {
    if (can(id)) return id;
  }
  return "settings";
}

export function hasPermission(user, key) {
  if (!user) return false;
  if (user.role === "admin" || (user.permissions || []).includes("*")) return true;
  return (user.permissions || []).includes(key);
}

export const ALL_ASSIGNABLE_PERMISSIONS = [
  { key: PERM.dashboard, label: "Dashboard" },
  { key: PERM.services, label: "Services" },
  { key: PERM.smoke, label: "Smoke tests" },
  { key: PERM.patching, label: "Manual testing" },
  { key: PERM.k8, label: "Kubernetes" },
  { key: PERM.ssl, label: "SSL" },
  { key: PERM.configTree, label: "Config tree" },
  { key: PERM.alerts, label: "Alerts" },
  { key: PERM.settingsAppearance, label: "Settings · appearance only" },
  { key: PERM.settingsFull, label: "Settings · full (API, runtime, audit, registry)" },
  { key: PERM.usersManage, label: "User management" },
  { key: PERM.configWrite, label: "Config write (services / patching / K8)" },
];
